File: sdk/go/request_boundary.go
package nextloggers

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"
)

// RequestTransport is a closed set of protocol families supported by the
// request failure boundary.
type RequestTransport string

const (
	RequestTransportHTTP      RequestTransport = "http"
	RequestTransportTCP       RequestTransport = "tcp"
	RequestTransportWebSocket RequestTransport = "websocket"
)

// RequestScope distinguishes a long-lived connection/session from one message.
type RequestScope string

const (
	RequestScopeRequest    RequestScope = "request"
	RequestScopeConnection RequestScope = "connection"
	RequestScopeSession    RequestScope = "session"
	RequestScopeMessage    RequestScope = "message"
)

// RequestFailureKind is intentionally closed so middleware adapters must map
// every recoverable outcome explicitly.
type RequestFailureKind string

const (
	RequestFailureException  RequestFailureKind = "exception"
	RequestFailurePanic      RequestFailureKind = "panic"
	RequestFailureTimeout    RequestFailureKind = "timeout"
	RequestFailureCancelled  RequestFailureKind = "cancelled"
	RequestFailureDisconnect RequestFailureKind = "disconnect"
)

// ErrPeerDisconnected can wrap a protocol-specific peer disconnect.
var ErrPeerDisconnected = errors.New("peer disconnected")

// RequestBoundary contains only allowlisted protocol correlation values. Never
// store payloads, authorization headers, cookies, raw tokens, or credentials.
type RequestBoundary struct {
	Transport    RequestTransport `json:"transport"`
	Scope        RequestScope     `json:"scope"`
	Phase        string           `json:"phase"`
	Operation    string           `json:"operation,omitempty"`
	ConnectionID string           `json:"connectionId,omitempty"`
	MessageID    string           `json:"messageId,omitempty"`
}

func HTTPRequestBoundary(phase, operation string) RequestBoundary {
	return RequestBoundary{
		Transport: RequestTransportHTTP,
		Scope:     RequestScopeRequest,
		Phase:     phase,
		Operation: operation,
	}
}

func TCPConnectionBoundary(phase, connectionID, operation string) RequestBoundary {
	return RequestBoundary{
		Transport:    RequestTransportTCP,
		Scope:        RequestScopeConnection,
		Phase:        phase,
		Operation:    operation,
		ConnectionID: connectionID,
	}
}

func TCPMessageBoundary(phase, connectionID, messageID, operation string) RequestBoundary {
	return RequestBoundary{
		Transport:    RequestTransportTCP,
		Scope:        RequestScopeMessage,
		Phase:        phase,
		Operation:    operation,
		ConnectionID: connectionID,
		MessageID:    messageID,
	}
}

func WebSocketSessionBoundary(phase, connectionID, operation string) RequestBoundary {
	return RequestBoundary{
		Transport:    RequestTransportWebSocket,
		Scope:        RequestScopeSession,
		Phase:        phase,
		Operation:    operation,
		ConnectionID: connectionID,
	}
}

func WebSocketMessageBoundary(phase, connectionID, messageID, operation string) RequestBoundary {
	return RequestBoundary{
		Transport:    RequestTransportWebSocket,
		Scope:        RequestScopeMessage,
		Phase:        phase,
		Operation:    operation,
		ConnectionID: connectionID,
		MessageID:    messageID,
	}
}

func (boundary RequestBoundary) normalized() (RequestBoundary, error) {
	boundary.Phase = strings.TrimSpace(boundary.Phase)
	boundary.Operation = strings.TrimSpace(boundary.Operation)
	boundary.ConnectionID = strings.TrimSpace(boundary.ConnectionID)
	boundary.MessageID = strings.TrimSpace(boundary.MessageID)
	if err := boundedBoundaryText("phase", boundary.Phase, 128, true); err != nil {
		return RequestBoundary{}, err
	}
	for field, value := range map[string]string{
		"operation":    boundary.Operation,
		"connectionId": boundary.ConnectionID,
		"messageId":    boundary.MessageID,
	} {
		if err := boundedBoundaryText(field, value, 256, false); err != nil {
			return RequestBoundary{}, err
		}
	}

	switch boundary.Transport {
	case RequestTransportHTTP:
		if boundary.Scope != RequestScopeRequest || boundary.ConnectionID != "" || boundary.MessageID != "" {
			return RequestBoundary{}, errors.New("HTTP boundaries require request scope and no connection/message IDs")
		}
	case RequestTransportTCP:
		switch boundary.Scope {
		case RequestScopeConnection:
			if boundary.MessageID != "" {
				return RequestBoundary{}, errors.New("TCP connection scope cannot carry a message ID")
			}
		case RequestScopeMessage:
		default:
			return RequestBoundary{}, errors.New("TCP boundaries require connection or message scope")
		}
	case RequestTransportWebSocket:
		switch boundary.Scope {
		case RequestScopeSession:
			if boundary.MessageID != "" {
				return RequestBoundary{}, errors.New("WebSocket session scope cannot carry a message ID")
			}
		case RequestScopeMessage:
		default:
			return RequestBoundary{}, errors.New("WebSocket boundaries require session or message scope")
		}
	default:
		return RequestBoundary{}, fmt.Errorf("unsupported request transport %q", boundary.Transport)
	}
	return boundary, nil
}

func boundedBoundaryText(field, value string, maximum int, required bool) error {
	if value == "" {
		if required {
			return fmt.Errorf("%s is required", field)
		}
		return nil
	}
	if len(value) > maximum || strings.ContainsAny(value, "\x00\r\n") {
		return fmt.Errorf("%s must be bounded text", field)
	}
	return nil
}

// RequestBoundaryFailure keeps the original error or recovered panic value and
// an immutable request-context snapshot for one protocol operation.
type RequestBoundaryFailure struct {
	Kind             RequestFailureKind `json:"kind"`
	Boundary         RequestBoundary    `json:"boundary"`
	Context          RequestContext     `json:"context"`
	Err              error              `json:"-"`
	Recovered        any                `json:"-"`
	ObservedAtUnixMS int64              `json:"observedAtUnixMs"`
}

// RequestBoundaryResult makes success and failure explicit without relying on
// a process-global panic or exception hook.
type RequestBoundaryResult[T any] struct {
	Value   T
	Failure *RequestBoundaryFailure
}

func (result RequestBoundaryResult[T]) OK() bool {
	return result.Failure == nil
}

// RequestBoundaryOptions controls classification and request-scoped reporting.
type RequestBoundaryOptions struct {
	Classify func(error, RequestBoundary) RequestFailureKind
	Report   func(context.Context, RequestBoundaryFailure)
	Now      func() time.Time
}

func defaultFailureKind(err error) RequestFailureKind {
	switch {
	case errors.Is(err, context.DeadlineExceeded):
		return RequestFailureTimeout
	case errors.Is(err, context.Canceled):
		return RequestFailureCancelled
	case errors.Is(err, ErrPeerDisconnected):
		return RequestFailureDisconnect
	default:
		return RequestFailureException
	}
}

func validFailureKind(kind RequestFailureKind) bool {
	switch kind {
	case RequestFailureException,
		RequestFailurePanic,
		RequestFailureTimeout,
		RequestFailureCancelled,
		RequestFailureDisconnect:
		return true
	default:
		return false
	}
}

func requestFailure(
	ctx context.Context,
	boundary RequestBoundary,
	kind RequestFailureKind,
	err error,
	recovered any,
	now func() time.Time,
) RequestBoundaryFailure {
	snapshot, ok := CaptureRequestContext(ctx)
	if !ok {
		snapshot = RequestContext{}
	}
	return RequestBoundaryFailure{
		Kind:             kind,
		Boundary:         boundary,
		Context:          snapshot,
		Err:              err,
		Recovered:        recovered,
		ObservedAtUnixMS: now().UnixMilli(),
	}
}

func reportRequestFailure(
	ctx context.Context,
	failure RequestBoundaryFailure,
	reporter func(context.Context, RequestBoundaryFailure),
) {
	if reporter == nil {
		return
	}
	defer func() {
		// Telemetry/reporting panics cannot replace the request's original result.
		_ = recover()
	}()
	reporter(ctx, failure)
}

// RunWithRequestBoundary installs the canonical context.Context value, catches
// errors and panics from exactly one logical HTTP/TCP/WebSocket operation, and
// reports while that context remains active. Child goroutines must still receive
// ctx explicitly and install their own recovery boundary.
func RunWithRequestBoundary[T any](
	parent context.Context,
	requestContext RequestContext,
	boundary RequestBoundary,
	operation func(context.Context) (T, error),
	options RequestBoundaryOptions,
) (result RequestBoundaryResult[T]) {
	if parent == nil {
		parent = context.Background()
	}
	ctx := WithRequestContext(parent, requestContext)
	now := options.Now
	if now == nil {
		now = time.Now
	}

	normalized, validationError := boundary.normalized()
	if validationError != nil {
		failure := requestFailure(
			ctx,
			boundary,
			RequestFailureException,
			validationError,
			nil,
			now,
		)
		reportRequestFailure(ctx, failure, options.Report)
		return RequestBoundaryResult[T]{Failure: &failure}
	}
	if operation == nil {
		failure := requestFailure(
			ctx,
			normalized,
			RequestFailureException,
			errors.New("request boundary operation is nil"),
			nil,
			now,
		)
		reportRequestFailure(ctx, failure, options.Report)
		return RequestBoundaryResult[T]{Failure: &failure}
	}

	defer func() {
		if recovered := recover(); recovered != nil {
			failure := requestFailure(
				ctx,
				normalized,
				RequestFailurePanic,
				nil,
				recovered,
				now,
			)
			reportRequestFailure(ctx, failure, options.Report)
			result = RequestBoundaryResult[T]{Failure: &failure}
		}
	}()

	value, err := operation(ctx)
	if err == nil {
		return RequestBoundaryResult[T]{Value: value}
	}
	kind := defaultFailureKind(err)
	if options.Classify != nil {
		func() {
			defer func() { _ = recover() }()
			candidate := options.Classify(err, normalized)
			if validFailureKind(candidate) {
				kind = candidate
			}
		}()
	}
	failure := requestFailure(ctx, normalized, kind, err, nil, now)
	reportRequestFailure(ctx, failure, options.Report)
	return RequestBoundaryResult[T]{Failure: &failure}
}

File: src/request-boundary.ts
import type { LogContext } from './base-logger.js';
import { cloneLogContext, type LogContextApi } from './context-shared.js';

export const requestFailureKinds = Object.freeze([
  'exception',
  'panic',
  'timeout',
  'cancelled',
  'disconnect',
] as const);

export type RequestFailureKind = (typeof requestFailureKinds)[number];

export type RequestBoundary =
  | Readonly<{
      transport: 'http';
      scope: 'request';
      phase: string;
      operation?: string;
    }>
  | Readonly<{
      transport: 'tcp';
      scope: 'connection' | 'message';
      phase: string;
      operation?: string;
      connectionId?: string;
      messageId?: string;
    }>
  | Readonly<{
      transport: 'websocket';
      scope: 'session' | 'message';
      phase: string;
      operation?: string;
      connectionId?: string;
      messageId?: string;
    }>;

export interface RequestBoundaryFailure {
  readonly kind: RequestFailureKind;
  readonly boundary: RequestBoundary;
  readonly context: Readonly<LogContext>;
  readonly cause: unknown;
  readonly observedAtUnixMs: number;
}

export type RequestBoundaryResult<T> =
  | Readonly<{ ok: true; value: T }>
  | Readonly<{ ok: false; failure: RequestBoundaryFailure }>;

export interface RequestBoundaryOptions {
  /** Classifies a caught value without changing the original failure cause. */
  classify?: (
    cause: unknown,
    boundary: RequestBoundary,
  ) => RequestFailureKind;
  /**
   * Runs before the context scope unwinds. Reporter failures are intentionally
   * ignored so telemetry cannot replace the request's original outcome.
   */
  report?: (
    failure: RequestBoundaryFailure,
  ) => void | Promise<void>;
  /** Injectable clock for deterministic conformance tests. */
  now?: () => number;
}

export interface RequestBoundaryApi {
  runWithRequestBoundary<T>(
    context: LogContext,
    boundary: RequestBoundary,
    operation: () => T | Promise<T>,
    options?: RequestBoundaryOptions,
  ): Promise<RequestBoundaryResult<T>>;
}

const validFailureKinds = new Set<RequestFailureKind>(requestFailureKinds);

function exhaustive(value: never): never {
  throw new TypeError(`unsupported request boundary: ${String(value)}`);
}

function boundedText(
  value: string,
  field: string,
  maximumLength: number,
): string {
  const normalized = value.trim();
  if (
    normalized.length === 0 ||
    normalized.length > maximumLength ||
    /[\u0000-\u001f\u007f]/u.test(normalized)
  ) {
    throw new TypeError(`${field} must be a non-empty bounded text value`);
  }
  return normalized;
}

function optionalText(
  value: string | undefined,
  field: string,
): string | undefined {
  return value === undefined ? undefined : boundedText(value, field, 256);
}

function normalizeBoundary(boundary: RequestBoundary): RequestBoundary {
  const phase = boundedText(boundary.phase, 'phase', 128);
  const operation = optionalText(boundary.operation, 'operation');

  switch (boundary.transport) {
    case 'http':
      return Object.freeze({
        transport: 'http',
        scope: 'request',
        phase,
        ...(operation === undefined ? {} : { operation }),
      });
    case 'tcp': {
      const connectionId = optionalText(boundary.connectionId, 'connectionId');
      const messageId = optionalText(boundary.messageId, 'messageId');
      return Object.freeze({
        transport: 'tcp',
        scope: boundary.scope,
        phase,
        ...(operation === undefined ? {} : { operation }),
        ...(connectionId === undefined ? {} : { connectionId }),
        ...(messageId === undefined ? {} : { messageId }),
      });
    }
    case 'websocket': {
      const connectionId = optionalText(boundary.connectionId, 'connectionId');
      const messageId = optionalText(boundary.messageId, 'messageId');
      return Object.freeze({
        transport: 'websocket',
        scope: boundary.scope,
        phase,
        ...(operation === undefined ? {} : { operation }),
        ...(connectionId === undefined ? {} : { connectionId }),
        ...(messageId === undefined ? {} : { messageId }),
      });
    }
    default:
      return exhaustive(boundary);
  }
}

function freezeContext(context: LogContext): Readonly<LogContext> {
  const snapshot = cloneLogContext(context);
  if (snapshot.loggedInUser !== undefined) Object.freeze(snapshot.loggedInUser);
  if (snapshot.users !== undefined) {
    for (const user of snapshot.users) Object.freeze(user);
    Object.freeze(snapshot.users);
  }
  if (snapshot.fields !== undefined) Object.freeze(snapshot.fields);
  if (snapshot.traceIds !== undefined) Object.freeze(snapshot.traceIds);
  if (snapshot.tags !== undefined) Object.freeze(snapshot.tags);
  return Object.freeze(snapshot);
}

function classifyFailure(
  cause: unknown,
  boundary: RequestBoundary,
  classify: RequestBoundaryOptions['classify'],
): RequestFailureKind {
  if (classify === undefined) return 'exception';
  try {
    const candidate = classify(cause, boundary);
    return validFailureKinds.has(candidate) ? candidate : 'exception';
  } catch {
    return 'exception';
  }
}

export function httpRequestBoundary(
  phase: string,
  operation?: string,
): RequestBoundary {
  return normalizeBoundary({
    transport: 'http',
    scope: 'request',
    phase,
    ...(operation === undefined ? {} : { operation }),
  });
}

export function tcpConnectionBoundary(
  phase: string,
  connectionId?: string,
  operation?: string,
): RequestBoundary {
  return normalizeBoundary({
    transport: 'tcp',
    scope: 'connection',
    phase,
    ...(connectionId === undefined ? {} : { connectionId }),
    ...(operation === undefined ? {} : { operation }),
  });
}

export function tcpMessageBoundary(
  phase: string,
  connectionId?: string,
  messageId?: string,
  operation?: string,
): RequestBoundary {
  return normalizeBoundary({
    transport: 'tcp',
    scope: 'message',
    phase,
    ...(connectionId === undefined ? {} : { connectionId }),
    ...(messageId === undefined ? {} : { messageId }),
    ...(operation === undefined ? {} : { operation }),
  });
}

export function webSocketSessionBoundary(
  phase: string,
  connectionId?: string,
  operation?: string,
): RequestBoundary {
  return normalizeBoundary({
    transport: 'websocket',
    scope: 'session',
    phase,
    ...(connectionId === undefined ? {} : { connectionId }),
    ...(operation === undefined ? {} : { operation }),
  });
}

export function webSocketMessageBoundary(
  phase: string,
  connectionId?: string,
  messageId?: string,
  operation?: string,
): RequestBoundary {
  return normalizeBoundary({
    transport: 'websocket',
    scope: 'message',
    phase,
    ...(connectionId === undefined ? {} : { connectionId }),
    ...(messageId === undefined ? {} : { messageId }),
    ...(operation === undefined ? {} : { operation }),
  });
}

/**
 * Builds one failure boundary over the runtime's canonical context carrier.
 * This helper catches only failures from the supplied logical operation. It
 * deliberately does not install process-global uncaught-exception handlers.
 */
export function createRequestBoundaryApi(
  contextApi: Pick<
    LogContextApi,
    'runWithLogContext' | 'captureLogContext'
  >,
): RequestBoundaryApi {
  return {
    async runWithRequestBoundary<T>(
      context: LogContext,
      boundary: RequestBoundary,
      operation: () => T | Promise<T>,
      options: RequestBoundaryOptions = {},
    ): Promise<RequestBoundaryResult<T>> {
      const explicitContext = freezeContext(context);
      const normalizedBoundary = normalizeBoundary(boundary);
      const now = options.now ?? Date.now;

      return contextApi.runWithLogContext(context, async () => {
        try {
          return Object.freeze({
            ok: true as const,
            value: await operation(),
          });
        } catch (cause) {
          const activeContext = contextApi.captureLogContext();
          const failure = Object.freeze({
            kind: classifyFailure(cause, normalizedBoundary, options.classify),
            boundary: normalizedBoundary,
            context:
              activeContext === undefined
                ? explicitContext
                : freezeContext(activeContext),
            cause,
            observedAtUnixMs: now(),
          }) satisfies RequestBoundaryFailure;

          if (options.report !== undefined) {
            try {
              await options.report(failure);
            } catch {
              // A logger/exporter failure must never replace the operation's
              // original failure or escape the protocol boundary.
            }
          }

          return Object.freeze({ ok: false as const, failure });
        }
      });
    },
  };
}
